### DISCUSSION TITLE
Bazel pre-releases are no longer in GitHub, any alternative?

### DISCUSSION THREAD

**Discussion by dragos** [bazelbuild/bazel#20421] [Q&A]
I noticed that very recently (over the weekend) all pre-release versions of Bazel are gone from https://github.com/bazelbuild/bazel/releases. Is there an alternative place to find releases for the upcoming 7.0 and 8.0 versions?

We run regression tests on upcoming versions to make sure we can upgrade, and this is something we'd like to keep doing. Are pre-releases still published somewhere else, or we need to build from source? I couldn't find any discussion about this, neither on bazel-user mailing list, github issues or discussions.

**Comment by dragos** [comment]
I found https://releases.bazel.build/rolling.html and, even though it's not json, it is easy enough to parse :)

**Comment by meteorcloudy** [comment]
Sorry for the breakage. Yes, we recently deleted rolling releases from the GitHub release page to make the more stable LTS releases more visible. We'll keep pushing pre-releases to https://releases.bazel.build/rolling.html and if you are using Bazelisk, just set `USE_BAZEL_VERSION=rolling`.

**Comment by srdo-humio** [comment]
The URLs on `releases.bazel.build` seem slightly less convenient for Bazelisk than the Github ones were, especially when running an internal pull through cache like Artifactory.

For Github releases, the prefix for all files is `https://github.com/bazelbuild/bazel/releases/download`, and Bazelisk is capable of appending the version and filename to get the right paths, e.g. `https://github.com/bazelbuild/bazel/releases/download/6.4.0/bazel-6.4.0-darwin-arm64`. This means we can easily mirror all releases by setting up a pull through cache that pulls from `https://github.com/bazelbuild/bazel/releases/download`, and pointing Bazelisk at the cache via `BAZELISK_BASE_URL`.

The path to one of the rolling prereleases now looks like `https://releases.bazel.build/8.0.0/rolling/8.0.0-pre.20231030.2/bazel-8.0.0-pre.20231030.2-darwin-arm64`. A part of the version string shows up in the middle of the url, which makes mirroring and Bazelisk configuration slightly harder. 

Would it be possible to make the same files available with a fixed prefix, e.g. `https://releases.bazel.build/all`, so the full path to a file would be similar to `https://releases.bazel.build/all/8.0.0-pre.20231030.2/bazel-8.0.0-pre.20231030.2-darwin-arm64`?

This would make mirroring as easy as it was for Github.

**Reply by meteorcloudy** [reply]
@srdo-humio That means we'll have to either duplicate those binaries in the new format or change Bazelisk to point to the new URL and that's a breaking change for all users. To mirror all rolling releases artifacts, can you just parse https://releases.bazel.build/rolling.html? I'm sure it won't take too long to write a script with some chatbot's help ;)

**Reply by srdo-humio** [reply]
I was hoping you could set up redirects/aliases instead, so e.g. `https://releases.bazel.build/all/8.0.0-pre.20231030.2/bazel-8.0.0-pre.20231030.2-darwin-arm64` points to `https://releases.bazel.build/8.0.0/rolling/8.0.0-pre.20231030.2/bazel-8.0.0-pre.20231030.2-darwin-arm64` (or the reverse, if that's easier), so there wouldn't be a need for duplication or breaking changes?

We can definitely make a script that mirrors the files, but fixing it on our end means it's still a problem for other people :) It was very convenient to not need a script, because it means a pull-through cache only needs to fetch the releases we actually use internally, and not all releases, and it can do so automatically in response to requests for a particular version.

**Reply by meteorcloudy** [reply]
Hmm, not sure how easy it is to set up redirects, maybe we'll have to do it for each major version? /cc @fweikert
